src/npc.cpp

```diff
@@ -642,6 +643,31 @@ void npc::revert_after_activity()
     current_activity_id = activity_id::NULL_ID();
     clear_destination();
     backlog.clear();
+    if( has_offscreen_job() ){
+        check_mission_resume();
+    }
+}
+
+bool npc::at_capacity_or_over_time_limit_for_travel_job()
+{
+    bool time_passed = travelling_work_started ? calendar::turn - *travelling_work_started > 6_hours : false;
+    if( volume_capacity() - volume_carried() < 750_ml || time_passed ){
+        return true;
+    }
+    return false;
+}
+
+void npc::check_mission_resume()
+{
+    if( offscreen_job == NPC_OFFSCREEN_JOB_FORAGE ){
```

Comment:
Let's do this right upfront.
Create an `npc_offscreen_job::` class which contains whatever generic data and functions all off-screen jobs need to have.
Create an `npc_offscreen_foraging::` class which is a child class of `npc_offscreen_job`.
Create an `npc_offscreen_no_job::` class which is a child class of `npc_offscreen_job` and is the default null job.
`if( has_offscreen_job() ) { check_mission_resume(); }` becomes `offscreen_job.work_on_job()` which is an immediate return in `npc_offscreen_no_job` and `if( at_capacity_or_over_time_list_for_travel_job() ) { ... ` for `npc_offscreen_foraging`

But basically we want to encapsulate all the details of how the offscreen job actually works in `npc_offscreen_job` and it's child classes.

Reply:
Well I've attempted that, my first ever go at this kinda structure, so it's probably all wrong, but its working, at least.